server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java

```diff
@@ -55,12 +62,18 @@ public boolean isWithinLeafObject() {
     }
 
     public String pathAsText(String name) {
-        sb.setLength(0);
-        for (int i = 0; i < index; i++) {
-            sb.append(path[i]).append(DELIMITER);
+        // If index is 0 we know that we are at the root, so return the provided string directly
+        if (index == 0) {
+            return name;
         }
-        sb.append(name);
-        return sb.toString();
+
+        // Otherwise we preserve the previously built StringBuilder and append any necessary path fields
+        while (sbIndex < index) {
```

Comment:
I think we can already keep the string builder current while we add / remove things and avoid the need to have a second index here